File: pkg/operations/file.go
package operations

import (
	"context"
	"fmt"
	"io"
	"os"
	"path/filepath"
)

// FileHandler handles file system operations
type FileHandler struct{}

// NewFileHandler creates a new file handler
func NewFileHandler() *FileHandler {
	return &FileHandler{}
}

// GetType returns the operation type
func (h *FileHandler) GetType() string {
	return "file"
}

// Validate checks if the parameters are valid
func (h *FileHandler) Validate(params map[string]interface{}) error {
	action, ok := params["action"].(string)
	if !ok {
		return fmt.Errorf("action parameter is required")
	}

	switch action {
	case "read", "delete", "exists":
		if _, ok := params["path"]; !ok {
			return fmt.Errorf("path parameter is required for %s action", action)
		}
	case "write", "append":
		if _, ok := params["path"]; !ok {
			return fmt.Errorf("path parameter is required for %s action", action)
		}
		if _, ok := params["content"]; !ok {
			return fmt.Errorf("content parameter is required for %s action", action)
		}
	case "copy":
		if _, ok := params["source"]; !ok {
			return fmt.Errorf("source parameter is required for copy action")
		}
		if _, ok := params["destination"]; !ok {
			return fmt.Errorf("destination parameter is required for copy action")
		}
	case "move", "rename":
		if _, ok := params["source"]; !ok {
			return fmt.Errorf("source parameter is required for %s action", action)
		}
		if _, ok := params["destination"]; !ok {
			return fmt.Errorf("destination parameter is required for %s action", action)
		}
	case "mkdir":
		if _, ok := params["path"]; !ok {
			return fmt.Errorf("path parameter is required for mkdir action")
		}
	case "list":
		if _, ok := params["path"]; !ok {
			return fmt.Errorf("path parameter is required for list action")
		}
	}

	return nil
}

// Execute performs the file operation
func (h *FileHandler) Execute(ctx context.Context, params map[string]interface{}, execCtx interface{}) (map[string]interface{}, error) {
	action := params["action"].(string)

	var result map[string]interface{}
	var err error

	switch action {
	case "read":
		result, err = h.executeRead(ctx, params)
	case "write":
		result, err = h.executeWrite(ctx, params)
	case "append":
		result, err = h.executeAppend(ctx, params)
	case "delete":
		result, err = h.executeDelete(ctx, params)
	case "copy":
		result, err = h.executeCopy(ctx, params)
	case "move":
		result, err = h.executeMove(ctx, params)
	case "rename":
		result, err = h.executeRename(ctx, params)
	case "exists":
		result, err = h.executeExists(ctx, params)
	case "mkdir":
		result, err = h.executeMkdir(ctx, params)
	case "list":
		result, err = h.executeList(ctx, params)
	default:
		return nil, fmt.Errorf("unsupported file action: %s", action)
	}

	return result, err
}

func (h *FileHandler) executeRead(ctx context.Context, params map[string]interface{}) (map[string]interface{}, error) {
	path := getStringParam(params, "path")

	content, err := os.ReadFile(path)
	if err != nil {
		return nil, fmt.Errorf("read failed: %v", err)
	}

	return map[string]interface{}{
		"success": true,
		"path":    path,
		"content": string(content),
		"size":    len(content),
	}, nil
}

func (h *FileHandler) executeWrite(ctx context.Context, params map[string]interface{}) (map[string]interface{}, error) {
	path := getStringParam(params, "path")
	content := getStringParam(params, "content")

	// Create directory if it doesn't exist
	dir := filepath.Dir(path)
	if err := os.MkdirAll(dir, 0755); err != nil {
		return nil, fmt.Errorf("create directory failed: %v", err)
	}

	err := os.WriteFile(path, []byte(content), 0644)
	if err != nil {
		return nil, fmt.Errorf("write failed: %v", err)
	}

	return map[string]interface{}{
		"success": true,
		"path":    path,
		"size":    len(content),
	}, nil
}

func (h *FileHandler) executeAppend(ctx context.Context, params map[string]interface{}) (map[string]interface{}, error) {
	path := getStringParam(params, "path")
	content := getStringParam(params, "content")

	file, err := os.OpenFile(path, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
	if err != nil {
		return nil, fmt.Errorf("open file failed: %v", err)
	}
	defer file.Close()

	if _, err := file.WriteString(content); err != nil {
		return nil, fmt.Errorf("append failed: %v", err)
	}

	return map[string]interface{}{
		"success":  true,
		"path":     path,
		"appended": len(content),
	}, nil
}

func (h *FileHandler) executeDelete(ctx context.Context, params map[string]interface{}) (map[string]interface{}, error) {
	path := getStringParam(params, "path")

	err := os.RemoveAll(path)
	if err != nil {
		return nil, fmt.Errorf("delete failed: %v", err)
	}

	return map[string]interface{}{
		"success": true,
		"path":    path,
	}, nil
}

func (h *FileHandler) executeCopy(ctx context.Context, params map[string]interface{}) (map[string]interface{}, error) {
	source := getStringParam(params, "source")
	destination := getStringParam(params, "destination")

	// Create destination directory if it doesn't exist
	dir := filepath.Dir(destination)
	if err := os.MkdirAll(dir, 0755); err != nil {
		return nil, fmt.Errorf("create destination directory failed: %v", err)
	}

	// Check if source is a directory
	info, err := os.Stat(source)
	if err != nil {
		return nil, fmt.Errorf("stat source failed: %v", err)
	}

	if info.IsDir() {
		err = copyDir(source, destination)
	} else {
		err = copyFile(source, destination)
	}

	if err != nil {
		return nil, fmt.Errorf("copy failed: %v", err)
	}

	return map[string]interface{}{
		"success":     true,
		"source":      source,
		"destination": destination,
	}, nil
}

func (h *FileHandler) executeMove(ctx context.Context, params map[string]interface{}) (map[string]interface{}, error) {
	source := getStringParam(params, "source")
	destination := getStringParam(params, "destination")

	// Create destination directory if it doesn't exist
	dir := filepath.Dir(destination)
	if err := os.MkdirAll(dir, 0755); err != nil {
		return nil, fmt.Errorf("create destination directory failed: %v", err)
	}

	err := os.Rename(source, destination)
	if err != nil {
		return nil, fmt.Errorf("move failed: %v", err)
	}

	return map[string]interface{}{
		"success":     true,
		"source":      source,
		"destination": destination,
	}, nil
}

func (h *FileHandler) executeRename(ctx context.Context, params map[string]interface{}) (map[string]interface{}, error) {
	source := getStringParam(params, "source")
	destination := getStringParam(params, "destination")

	err := os.Rename(source, destination)
	if err != nil {
		return nil, fmt.Errorf("rename failed: %v", err)
	}

	return map[string]interface{}{
		"success":     true,
		"source":      source,
		"destination": destination,
	}, nil
}

func (h *FileHandler) executeExists(ctx context.Context, params map[string]interface{}) (map[string]interface{}, error) {
	path := getStringParam(params, "path")

	_, err := os.Stat(path)
	exists := err == nil

	return map[string]interface{}{
		"success": true,
		"path":    path,
		"exists":  exists,
	}, nil
}

func (h *FileHandler) executeMkdir(ctx context.Context, params map[string]interface{}) (map[string]interface{}, error) {
	path := getStringParam(params, "path")
	recursive := getBoolParamDefault(params, "recursive", true)

	var err error
	if recursive {
		err = os.MkdirAll(path, 0755)
	} else {
		err = os.Mkdir(path, 0755)
	}

	if err != nil {
		return nil, fmt.Errorf("mkdir failed: %v", err)
	}

	return map[string]interface{}{
		"success": true,
		"path":    path,
	}, nil
}

func (h *FileHandler) executeList(ctx context.Context, params map[string]interface{}) (map[string]interface{}, error) {
	path := getStringParam(params, "path")

	entries, err := os.ReadDir(path)
	if err != nil {
		return nil, fmt.Errorf("list failed: %v", err)
	}

	files := []map[string]interface{}{}
	for _, entry := range entries {
		info, _ := entry.Info()
		files = append(files, map[string]interface{}{
			"name":   entry.Name(),
			"is_dir": entry.IsDir(),
			"size":   info.Size(),
		})
	}

	return map[string]interface{}{
		"success": true,
		"path":    path,
		"files":   files,
		"count":   len(files),
	}, nil
}

// Helper functions
func copyFile(src, dst string) error {
	sourceFile, err := os.Open(src)
	if err != nil {
		return err
	}
	defer sourceFile.Close()

	destFile, err := os.Create(dst)
	if err != nil {
		return err
	}
	defer destFile.Close()

	if _, err := io.Copy(destFile, sourceFile); err != nil {
		return err
	}

	sourceInfo, err := os.Stat(src)
	if err != nil {
		return err
	}

	return os.Chmod(dst, sourceInfo.Mode())
}

func copyDir(src, dst string) error {
	sourceInfo, err := os.Stat(src)
	if err != nil {
		return err
	}

	err = os.MkdirAll(dst, sourceInfo.Mode())
	if err != nil {
		return err
	}

	entries, err := os.ReadDir(src)
	if err != nil {
		return err
	}

	for _, entry := range entries {
		srcPath := filepath.Join(src, entry.Name())
		dstPath := filepath.Join(dst, entry.Name())

		if entry.IsDir() {
			err = copyDir(srcPath, dstPath)
		} else {
			err = copyFile(srcPath, dstPath)
		}

		if err != nil {
			return err
		}
	}

	return nil
}

func getBoolParamDefault(params map[string]interface{}, key string, defaultValue bool) bool {
	if val, ok := params[key]; ok {
		if b, ok := val.(bool); ok {
			return b
		}
	}
	return defaultValue
}

File: src/operations/file.py
"""
File system operations handler for the auto-workflow system.
"""
import os
import shutil
from typing import Dict, Any, List
from .handler import Handler
from ..models import ExecutionContext


class FileHandler(Handler):
    """Handles file system operations."""

    def get_type(self) -> str:
        """Return the operation type."""
        return "file"

    def validate(self, params: Dict[str, Any]) -> None:
        """Check if the parameters are valid."""
        action = params.get("action")
        if not action or not isinstance(action, str):
            raise ValueError("action parameter is required")

        if action in ["read", "delete", "exists"]:
            if "path" not in params:
                raise ValueError(f"path parameter is required for {action} action")
        elif action in ["write", "append"]:
            if "path" not in params:
                raise ValueError(f"path parameter is required for {action} action")
            if "content" not in params:
                raise ValueError(f"content parameter is required for {action} action")
        elif action == "copy":
            if "source" not in params:
                raise ValueError("source parameter is required for copy action")
            if "destination" not in params:
                raise ValueError("destination parameter is required for copy action")
        elif action in ["move", "rename"]:
            if "source" not in params:
                raise ValueError(f"source parameter is required for {action} action")
            if "destination" not in params:
                raise ValueError(f"destination parameter is required for {action} action")
        elif action == "mkdir":
            if "path" not in params:
                raise ValueError("path parameter is required for mkdir action")
        elif action == "list":
            if "path" not in params:
                raise ValueError("path parameter is required for list action")

    def execute(self, params: Dict[str, Any], exec_ctx: ExecutionContext) -> Dict[str, Any]:
        """Perform the file operation."""
        action = params["action"]

        if action == "read":
            return self._execute_read(params)
        elif action == "write":
            return self._execute_write(params)
        elif action == "append":
            return self._execute_append(params)
        elif action == "delete":
            return self._execute_delete(params)
        elif action == "copy":
            return self._execute_copy(params)
        elif action == "move":
            return self._execute_move(params)
        elif action == "rename":
            return self._execute_rename(params)
        elif action == "exists":
            return self._execute_exists(params)
        elif action == "mkdir":
            return self._execute_mkdir(params)
        elif action == "list":
            return self._execute_list(params)
        else:
            raise ValueError(f"unsupported file action: {action}")

    def _execute_read(self, params: Dict[str, Any]) -> Dict[str, Any]:
        """Execute read operation."""
        path = self._get_string_param(params, "path")

        with open(path, 'r', encoding='utf-8') as f:
            content = f.read()

        return {
            "success": True,
            "path": path,
            "content": content,
            "size": len(content),
        }

    def _execute_write(self, params: Dict[str, Any]) -> Dict[str, Any]:
        """Execute write operation."""
        path = self._get_string_param(params, "path")
        content = self._get_string_param(params, "content")

        # Create directory if it doesn't exist
        dir_path = os.path.dirname(path)
        if dir_path:
            os.makedirs(dir_path, exist_ok=True)

        with open(path, 'w', encoding='utf-8') as f:
            f.write(content)

        return {
            "success": True,
            "path": path,
            "size": len(content),
        }

    def _execute_append(self, params: Dict[str, Any]) -> Dict[str, Any]:
        """Execute append operation."""
        path = self._get_string_param(params, "path")
        content = self._get_string_param(params, "content")

        with open(path, 'a', encoding='utf-8') as f:
            f.write(content)

        return {
            "success": True,
            "path": path,
            "appended": len(content),
        }

    def _execute_delete(self, params: Dict[str, Any]) -> Dict[str, Any]:
        """Execute delete operation."""
        path = self._get_string_param(params, "path")

        if os.path.isdir(path):
            shutil.rmtree(path)
        else:
            os.remove(path)

        return {
            "success": True,
            "path": path,
        }

    def _execute_copy(self, params: Dict[str, Any]) -> Dict[str, Any]:
        """Execute copy operation."""
        source = self._get_string_param(params, "source")
        destination = self._get_string_param(params, "destination")

        # Create destination directory if it doesn't exist
        dir_path = os.path.dirname(destination)
        if dir_path:
            os.makedirs(dir_path, exist_ok=True)

        # Check if source is a directory
        if os.path.isdir(source):
            shutil.copytree(source, destination)
        else:
            shutil.copy2(source, destination)

        return {
            "success": True,
            "source": source,
            "destination": destination,
        }

    def _execute_move(self, params: Dict[str, Any]) -> Dict[str, Any]:
        """Execute move operation."""
        source = self._get_string_param(params, "source")
        destination = self._get_string_param(params, "destination")

        # Create destination directory if it doesn't exist
        dir_path = os.path.dirname(destination)
        if dir_path:
            os.makedirs(dir_path, exist_ok=True)

        shutil.move(source, destination)

        return {
            "success": True,
            "source": source,
            "destination": destination,
        }

    def _execute_rename(self, params: Dict[str, Any]) -> Dict[str, Any]:
        """Execute rename operation."""
        source = self._get_string_param(params, "source")
        destination = self._get_string_param(params, "destination")

        os.rename(source, destination)

        return {
            "success": True,
            "source": source,
            "destination": destination,
        }

    def _execute_exists(self, params: Dict[str, Any]) -> Dict[str, Any]:
        """Execute exists operation."""
        path = self._get_string_param(params, "path")

        exists = os.path.exists(path)

        return {
            "success": True,
            "path": path,
            "exists": exists,
        }

    def _execute_mkdir(self, params: Dict[str, Any]) -> Dict[str, Any]:
        """Execute mkdir operation."""
        path = self._get_string_param(params, "path")
        recursive = self._get_bool_param(params, "recursive", True)

        if recursive:
            os.makedirs(path, exist_ok=True)
        else:
            os.makedirs(path, exist_ok=False)

        return {
            "success": True,
            "path": path,
        }

    def _execute_list(self, params: Dict[str, Any]) -> Dict[str, Any]:
        """Execute list operation."""
        path = self._get_string_param(params, "path")

        entries = os.listdir(path)
        files = []

        for entry in entries:
            entry_path = os.path.join(path, entry)
            is_dir = os.path.isdir(entry_path)
            size = os.path.getsize(entry_path) if not is_dir else 0

            files.append({
                "name": entry,
                "is_dir": is_dir,
                "size": size,
            })

        return {
            "success": True,
            "path": path,
            "files": files,
            "count": len(files),
        }

    # Helper methods
    def _get_string_param(self, params: Dict[str, Any], key: str) -> str:
        """Get string parameter."""
        value = params.get(key)
        if isinstance(value, str):
            return value
        return str(value) if value is not None else ""

    def _get_bool_param(self, params: Dict[str, Any], key: str, default: bool = False) -> bool:
        """Get boolean parameter."""
        value = params.get(key)
        if isinstance(value, bool):
            return value
        return default
